client/my-sites/patterns/home/index.tsx

```diff
@@ -34,8 +54,54 @@ export const PatternsHomePage = ( {
 	const locale = useLocale();
 	const isLoggedIn = useSelector( isUserLoggedIn );
 
+	const [ searchTerm, setSearchTerm ] = useState( () => {
+		if ( typeof window !== 'undefined' ) {
+			const params = new URLSearchParams( window.location.search );
+			return params.get( 's' ) ?? '';
+		}
+
+		return '';
+	} );
+
+	const isPatternsDisplaying = Boolean( category || searchTerm );
+
 	const { data: categories } = usePatternCategories( locale );
-	const { data: patterns } = usePatterns( locale, category );
+	const { data: patterns = [] } = usePatterns( locale, category, {
+		enabled: isPatternsDisplaying,
+	} );
+
+	const patternSearchResults = filterPatternsByQuery( patterns, searchTerm );
+	const patternsToRender = searchTerm ? patternSearchResults : patterns;
+
+	useEffect( () => {
+		setSearchTerm( '' );
+	}, [ category ] );
+
+	useEffect( () => {
+		const params = new URLSearchParams( window.location.search );
+
+		if ( searchTerm ) {
+			params.set( 's', searchTerm );
+		} else {
+			params.delete( 's' );
+		}
+
+		const paramsString = params.toString().length ? `?${ params.toString() }` : '';
+		page( `${ window.location.pathname }${ paramsString }` );
+	}, [ searchTerm ] );
+
+	useEffect( () => {
+		function onPopstate() {
+			const params = new URLSearchParams( window.location.search );
+			setSearchTerm( params.get( 's' ) ?? '' );
+		}
+
+		window.addEventListener( 'popstate', onPopstate );
+
+		return () => {
+			window.removeEventListener( 'popstate', onPopstate );
+		};
+	}, [] );
```

Comment:
It wasn't immediately clear to me what those effects are supposed to do, maybe this could be improved with comments?

```suggestion
	// Resets the search term whenever the category changes
	useEffect( () => {
		setSearchTerm( '' );
	}, [ category ] );

	// Updates the url of the page whenever the search term changes
	useEffect( () => {
		const params = new URLSearchParams( window.location.search );

		if ( searchTerm ) {
			params.set( 's', searchTerm );
		} else {
			params.delete( 's' );
		}

		const paramsString = params.toString().length ? `?${ params.toString() }` : '';
		page( `${ window.location.pathname }${ paramsString }` );
	}, [ searchTerm ] );

	// Updates the search term whenever the url of the page changes
	useEffect( () => {
		function onPopstate() {
			const params = new URLSearchParams( window.location.search );
			setSearchTerm( params.get( 's' ) ?? '' );
		}

		window.addEventListener( 'popstate', onPopstate );

		return () => {
			window.removeEventListener( 'popstate', onPopstate );
		};
	}, [] );
```